Microsoft.Toolkit.Uwp.UI.Controls/WrapPanel/WrapPanel.cs

```diff
@@ -0,0 +1,157 @@
+
+using System;
+using System.Collections.Generic;
+using Windows.Foundation;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls;
+
+namespace Microsoft.Toolkit.Uwp.UI.Controls.WrapPanel
+{
+    /// <summary>
+    /// WrapPanel is a panel that position child control vertically or horizontally based on the orientation and when max width/ max height is recieved a new row(in case of horizontal) or column (in case of vertical) is created to fit new controls.
+    /// </summary>
+    public partial class WrapPanel : Panel
+    {
+        private List<double> _vSize;
+
+        /// <summary>
+        /// Gets or sets the orientation of the WrapPanel, Horizontal or vertical means that child controls will be added horizontally until the width of the panel can't fit more control then a new row is added to fit new horizontal added child controls, vertical means that child will be added vertically until the height of the panel is recieved then a new column is added
+        /// </summary>
+        public Orientation Orientation
+        {
+            get { return (Orientation)GetValue(OrientationProperty); }
+            set { SetValue(OrientationProperty, value); }
+        }
+
+        /// <summary>
+        /// Identifies the <see cref="Orientation"/> dependency property.
+        /// </summary>
+        public static readonly DependencyProperty OrientationProperty =
+            DependencyProperty.Register(
+                "Orientation",
+                typeof(Orientation),
+                typeof(WrapPanel),
+                new PropertyMetadata(Orientation.Horizontal));
+
+        /// <inheritdoc />
+        protected override Size MeasureOverride(Size availableSize)
+        {
+            _vSize = new List<double> { 0.0 };
+            var totalMeasure = new UvMeasure();
+            var parentMeasure = new UvMeasure(Orientation, availableSize.Width, availableSize.Height);
+            var lineMeasure = new UvMeasure();
+            foreach (var child in Children)
+            {
+                child.Measure(availableSize);
+
+                var currentMeasure = new UvMeasure(Orientation, child.DesiredSize.Width, child.DesiredSize.Height);
+
+                if (parentMeasure.U > currentMeasure.U + lineMeasure.U)
+                {
+                    lineMeasure.U += currentMeasure.U;
+                    lineMeasure.V = Math.Max(lineMeasure.V, currentMeasure.V);
+                }
+                else
+                {
+                    // new line should be added
+                    // to get the max U to provide it correctly to ui width ex: ---| or -----|
+                    totalMeasure.U = Math.Max(lineMeasure.U, totalMeasure.U);
+                    totalMeasure.V += lineMeasure.V;
+                    _vSize.Add(lineMeasure.V);
+
+                    // if the next new row still can handle more controls
+                    if (parentMeasure.U > currentMeasure.U)
```

Comment:
This check is needed because when adding a new line we still didn't add the control to this line and we have 2 cases:
1) the control is taking the whole new line so another new line should be added (The else condition).
2) the control can be added to this line without taking the full width so we still can add another controls to this line in the new loop.

so that in the first case (the else) we will add a new blank empty line for the next control and in the second case we set the line measure to the current measure so that the next control start after the current measure.